server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java

```diff
@@ -0,0 +1,485 @@
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterState.VotingConfiguration;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * The core class of the cluster state coordination algorithm, directly implementing the
+ * <a href="https://github.com/elastic/elasticsearch-formal-models/blob/master/ZenWithTerms/tla/ZenWithTerms.tla">formal model</a>
+ */
+public class CoordinationState extends AbstractComponent {
+
+    private final DiscoveryNode localNode;
+
+    // persisted state
+    private final PersistedState persistedState;
+
+    // transient state
+    private VoteCollection joinVotes;
+    private boolean startedJoinSinceLastReboot;
+    private boolean electionWon;
+    private long lastPublishedVersion;
+    private VotingConfiguration lastPublishedConfiguration;
+    private VoteCollection publishVotes;
+
+    public CoordinationState(Settings settings, DiscoveryNode localNode, PersistedState persistedState) {
+        super(settings);
+
+        this.localNode = localNode;
+
+        // persisted state
+        this.persistedState = persistedState;
+
+        // transient state
+        this.joinVotes = new VoteCollection();
+        this.startedJoinSinceLastReboot = false;
+        this.electionWon = false;
+        this.lastPublishedVersion = 0L;
+        this.lastPublishedConfiguration = persistedState.getLastAcceptedState().getLastAcceptedConfiguration();
+        this.publishVotes = new VoteCollection();
+    }
+
+    public long getCurrentTerm() {
+        return persistedState.getCurrentTerm();
+    }
+
+    public ClusterState getLastAcceptedState() {
+        return persistedState.getLastAcceptedState();
+    }
+
+    public long getLastAcceptedTerm() {
+        return getLastAcceptedState().term();
+    }
+
+    public long getLastAcceptedVersion() {
+        return getLastAcceptedState().version();
+    }
+
+    public VotingConfiguration getLastCommittedConfiguration() {
+        return getLastAcceptedState().getLastCommittedConfiguration();
+    }
+
+    public VotingConfiguration getLastAcceptedConfiguration() {
+        return getLastAcceptedState().getLastAcceptedConfiguration();
+    }
+
+    public long getLastPublishedVersion() {
+        return lastPublishedVersion;
+    }
+
+    public boolean electionWon() {
+        return electionWon;
+    }
+
+    public boolean isElectionQuorum(VoteCollection votes) {
+        return votes.isQuorum(getLastCommittedConfiguration()) && votes.isQuorum(getLastAcceptedConfiguration());
+    }
+
+    public boolean isPublishQuorum(VoteCollection votes) {
+        return votes.isQuorum(getLastCommittedConfiguration()) && votes.isQuorum(lastPublishedConfiguration);
```

Comment:
Tests pass with
- `return votes.isQuorum(getLastCommittedConfiguration());`
- `return votes.isQuorum(lastPublishedConfiguration);`

Reply:
fixed in 8ee68b9d79a